File: PanamaJavaExamples/src/mscs/examples/RouteLoopbackTest.java
// Copyright © 2026 Khrustal & Mann
//              MELBOURNE, VICTORIA, AUSTRALIA, 3000
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied. See the License for the specific language governing
// permissions and limitations under the License.
//
// RouteLoopbackTest.java  (Java -- Panama FFI over TargetFacade)
//
// FOUR HUBS IN A TWO-LEVEL TREE, one process, no OS handles -- and the question
// is whether a message finds its way between two hubs that are NOT directly
// connected.
//
//              Root
//             /    \
//            A      B
//            |
//          Leaf
//
// READ THIS FIRST: THE ORIGINAL IS NOT AN MSCS HARNESS. Every other example in
// DirectExamples drives Targetcore. RouteLoopbackTest does not: it exercises
// treehub_runtime's PeerNetwork::route(), the clean-room in-process router the
// code generator emits generated projects against, with no P2PeerHub, no
// PostP2Pmsg and no pump thread anywhere. Its own header says so.
//
// So there is no facade translation of that code -- there is nothing of
// Targetcore in it to put a facade over. What this file does instead, following
// the Light tree, is ask the original's QUESTION of the real kernel: build the
// same four-node tree out of facade hubs, wire the same parent<->child edges
// over Dmx (the in-process transport, closest to route()'s DIRECT
// copy-on-deliver), and see how MSCS itself handles the scenarios.
//
// The addresses are hierarchical and dotted, which is what makes this a tree
// rather than four unrelated peers: "Root.A.Leaf" is a child of "Root.A" is a
// child of "Root". Routing and broadcast relay both key off that.
//
//   [1] Leaf -> B    : non-adjacent. Up to the root, then down.
//   [2] Root -> Leaf : straight down the tree, one intermediate hop.
//   [3] Leaf -> Z    : nobody by that address -- must NOT be delivered anywhere.
//   [4] Root broadcast : relayed down to every descendant. No counterpart in the
//                        original; it is the tree-shaped behaviour MSCS has and
//                        the generated router does not.
//
// THE ONE THING THIS FILE HAS TO DO THAT THE C++ VERSION DOES NOT. Four hubs
// means four pump threads calling into four Java sinks at once, so the delivery
// counters really are shared mutable state across threads -- the C++ version
// reaches for InterlockedIncrement for exactly this reason and this one reaches
// for AtomicInteger. Scenario [3] is the one that depends on it: it proves a
// negative by summing all four counters before and after, and a lost increment
// would turn a routing bug into a PASS.
//
// Verdict by EXIT CODE:
//   0 = SUCCESS : all four scenarios behaved as expected.
//   3 = FAIL    : a message was delivered somewhere it should not have been, or
//                 failed to arrive where it should have.
//   1 = SETUP   : the mesh could not be built.
package mscs.examples;

import java.util.concurrent.atomic.AtomicInteger;

import mscs.harness.Harness;
import mscs.p2pf.Abi;
import mscs.p2pf.Hub;
import mscs.p2pf.Network;

public final class RouteLoopbackTest {

    private static final String ROOT = "RouteMesh.Root";
    private static final String A    = "RouteMesh.Root.A";
    private static final String B    = "RouteMesh.Root.B";
    private static final String LEAF = "RouteMesh.Root.A.Leaf";
    private static final String Z    = "RouteMesh.Root.Z";      // nobody

    private static final String SVC_ROOT_A = "RM_RootA_Java";
    private static final String SVC_ROOT_B = "RM_RootB_Java";
    private static final String SVC_A_LEAF = "RM_ALeaf_Java";

    private static final String TOPIC = "route";

    /** One recorder per node: what arrived, and from where. */
    private static final class Node {
        final String        name;
        final AtomicInteger count = new AtomicInteger();
        final Harness.Gate  gate  = new Harness.Gate();
        Node(String name) { this.name = name; }
    }

    private static void attach(Hub hub, Node node) {
        hub.onTopic(TOPIC, m -> {
            node.count.incrementAndGet();
            Harness.banner(String.format("    [deliver] %s received \"%s\" (src=%s%s)",
                    node.name, m.text(), m.source, m.broadcast ? ", broadcast" : ""));
            node.gate.open();
        });
        hub.onError(w -> Harness.banner("    [error]   at " + node.name + ": " + w));
    }

    private static boolean check(String label, boolean ok) {
        Harness.banner("    => " + (ok ? "PASS" : "FAIL") + ": " + label);
        Harness.banner("");
        return ok;
    }

    public static void main(String[] args) {
        Harness.run("RouteLoopbackTest", () -> {
            Harness.banner("=== RouteLoopbackTest (Java) - four hubs, one process, tree routing ===");
            Harness.banner("        " + ROOT);
            Harness.banner("       /        \\");
            Harness.banner("    " + A + "   " + B);
            Harness.banner("      |");
            Harness.banner("   " + LEAF);
            Harness.banner("");

            Node nRoot = new Node(ROOT), nA = new Node(A), nB = new Node(B), nLeaf = new Node(LEAF);

            try (Network net = Network.open()) {
                Harness.log("MAIN", "facade : %s", net.versionString());

                // ---- Build the tree. Listeners first: dmx:// dials do not retry. ----
                //
                // Every edge here joins a parent to a child, which is what makes
                // the routing this harness measures possible at all: the
                // kernel's onward rules require an ancestor/descendant relation,
                // so a sibling edge would carry direct traffic and still refuse
                // to be a transit hop. The arming verbs classify each pair and
                // would say so -- every call below returns a plain S_OK.
                String epRootA = Abi.dmx(SVC_ROOT_A);
                String epRootB = Abi.dmx(SVC_ROOT_B);
                String epALeaf = Abi.dmx(SVC_A_LEAF);

                try (Hub root = net.createHub(ROOT);
                     Hub a    = net.createHub(A);
                     Hub b    = net.createHub(B);
                     Hub leaf = net.createHub(LEAF)) {

                    attach(root, nRoot);
                    int hr = root.listen(A, epRootA);
                    if (!Abi.failed(hr)) hr = root.listen(B, epRootB);
                    if (Abi.failed(hr)) {
                        Harness.log("MAIN", "FATAL: root Dmx listen failed (%s)", Abi.name(hr));
                        return Harness.EXIT_SETUP;
                    }

                    attach(a, nA);
                    hr = a.listen(LEAF, epALeaf);           // arm before Leaf dials
                    if (!Abi.failed(hr)) hr = a.connect(ROOT, epRootA);
                    if (Abi.failed(hr)) {
                        Harness.log("MAIN", "FATAL: A wiring failed (%s)", Abi.name(hr));
                        return Harness.EXIT_SETUP;
                    }

                    attach(b, nB);
                    hr = b.connect(ROOT, epRootB);
                    if (Abi.failed(hr)) {
                        Harness.log("MAIN", "FATAL: B wiring failed (%s)", Abi.name(hr));
                        return Harness.EXIT_SETUP;
                    }

                    attach(leaf, nLeaf);
                    hr = leaf.connect(A, epALeaf);
                    if (Abi.failed(hr)) {
                        Harness.log("MAIN", "FATAL: Leaf wiring failed (%s)", Abi.name(hr));
                        return Harness.EXIT_SETUP;
                    }

                    // Give the three links their login handshakes.
                    Harness.sleep(1_500);
                    Harness.banner("Links: Root<->A " + (root.isPeerUp(A)   ? "up" : "DOWN")
                                 + "   Root<->B "     + (root.isPeerUp(B)   ? "up" : "DOWN")
                                 + "   A<->Leaf "     + (a.isPeerUp(LEAF)   ? "up" : "DOWN"));
                    Harness.banner("");

                    boolean allOk = true;

                    // ---- [1] Non-adjacent, up then down: Leaf -> B ----------
                    Harness.banner("[1] Leaf -> " + B + "  (non-adjacent: up to the root, then down)");
                    hr = leaf.sendText(B, TOPIC, "hello from the leaf");
                    Harness.banner("    sendText returned " + Abi.name(hr));
                    boolean got1 = nB.gate.await(5_000);
                    allOk &= check("Leaf -> B delivered across two hops", got1 && nB.count.get() == 1);

                    // ---- [2] Straight down: Root -> Leaf --------------------
                    Harness.banner("[2] Root -> " + LEAF + "  (straight down the tree)");
                    hr = root.sendText(LEAF, TOPIC, "ping, down we go");
                    Harness.banner("    sendText returned " + Abi.name(hr));
                    boolean got2 = nLeaf.gate.await(5_000);
                    allOk &= check("Root -> Leaf delivered via A", got2 && nLeaf.count.get() == 1);

                    // ---- [3] No such address -------------------------------
                    Harness.banner("[3] Leaf -> " + Z + "  (unknown branch; must not be delivered)");
                    int before = nRoot.count.get() + nA.count.get()
                               + nB.count.get()   + nLeaf.count.get();
                    hr = leaf.sendText(Z, TOPIC, "lost, no such node");
                    Harness.banner("    sendText returned " + Abi.name(hr));
                    Harness.sleep(1_500);
                    int after = nRoot.count.get() + nA.count.get()
                              + nB.count.get()   + nLeaf.count.get();
                    allOk &= check("Leaf -> Z delivered to nobody", before == after);

                    // ---- [4] Broadcast relayed down the tree ----------------
                    Harness.banner("[4] Root broadcast  (relayed to every descendant -- no counterpart");
                    Harness.banner("    in the original, this is the tree behaviour MSCS adds)");
                    int aBefore = nA.count.get(), bBefore = nB.count.get(),
                        leafBefore = nLeaf.count.get();
                    hr = root.broadcastText(TOPIC, "all hands");
                    Harness.banner("    broadcast returned " + Abi.name(hr));
                    Harness.sleep(2_000);
                    Harness.banner("    reached: A " + (nA.count.get()    > aBefore    ? "yes" : "no")
                                 + "   B "           + (nB.count.get()    > bBefore    ? "yes" : "no")
                                 + "   Leaf "        + (nLeaf.count.get() > leafBefore ? "yes" : "no"));
                    allOk &= check("broadcast reached both direct children",
                                   nA.count.get() > aBefore && nB.count.get() > bBefore);

                    Harness.log("MAIN", "shutdown begin");
                    return Harness.verdict(allOk,
                            "the tree routed every scenario as expected",
                            "a route did not match expectations");
                }
            }
        });
    }
}

File: FacadeExamples/RouteLoopbackTest/RouteLoopbackTest.cpp
// Copyright © 2026 Khrustal & Mann
//              MELBOURNE, VICTORIA, AUSTRALIA, 3000
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied. See the License for the specific language governing
// permissions and limitations under the License.
//
// RouteLoopbackTest.cpp  (Light -- TargetFacade)
//
// FOUR HUBS IN A TWO-LEVEL TREE, one process, no OS handles -- and the question
// is whether a message finds its way between two hubs that are NOT directly
// connected.
//
//              Root
//             /    \
//            A      B
//            |
//          Leaf
//
// READ THIS FIRST: THE ORIGINAL IS NOT AN MSCS HARNESS. Every other example in
// DirectExamples drives Targetcore. RouteLoopbackTest does not: it exercises
// treehub_runtime's PeerNetwork::route(), the clean-room in-process router the
// code generator emits generated projects against, with no P2PeerHub, no
// PostP2Pmsg and no pump thread anywhere. Its own header says so.
//
// So there is no facade translation of that code -- there is nothing of
// Targetcore in it to put a facade over. What this file does instead is ask the
// original's QUESTION of the real kernel: build the same four-node tree out of
// facade hubs, wire the same parent<->child edges (over Dmx, the in-process
// transport -- the closest thing to route()'s DIRECT copy-on-deliver), and see
// how MSCS itself handles the three scenarios.
//
// The addresses are hierarchical and dotted, which is what makes this a tree
// rather than four unrelated peers: "Root.A.Leaf" is a child of "Root.A" is a
// child of "Root". Routing and broadcast relay both key off that.
//
//   [1] Leaf -> B    : non-adjacent. Up to the root, then down.
//   [2] Root -> Leaf : straight down the tree, one intermediate hop.
//   [3] Leaf -> Z    : nobody by that address -- must NOT be delivered anywhere.
//   [4] Root broadcast : relayed down to every descendant. This one has no
//                        counterpart in the original; it is the tree-shaped
//                        behaviour MSCS has and the generated router does not.
//
// Verdict by EXIT CODE:
//   0 = SUCCESS : all four scenarios behaved as expected.
//   3 = FAIL    : a message was delivered somewhere it should not have been,
//                 or failed to arrive where it should have.
//   1 = SETUP   : the mesh could not be built.

#include "TargetFacadeFn.hpp"
#include "LightHarness.h"

static const wchar_t* const kRoot = L"RouteMesh.Root";
static const wchar_t* const kA    = L"RouteMesh.Root.A";
static const wchar_t* const kB    = L"RouteMesh.Root.B";
static const wchar_t* const kLeaf = L"RouteMesh.Root.A.Leaf";
static const wchar_t* const kZ    = L"RouteMesh.Root.Z";      // nobody

static const wchar_t* const kSvcRootA = L"RM_RootA";
static const wchar_t* const kSvcRootB = L"RM_RootB";
static const wchar_t* const kSvcALeaf = L"RM_ALeaf";

static const wchar_t* const kTopic = L"route";

// One recorder per node: what arrived, and from where.
struct Node
{
    const wchar_t *name;
    volatile LONG  count;
    light::Gate    gate;

    Node ( const wchar_t *n ) : name ( n ), count ( 0 ) { }
};

static void Attach ( p2pf::Hub& hub, Node& node )
{
    hub.onTopic ( kTopic, [&node] ( const p2pf::Message& m )
    {
        ::InterlockedIncrement ( &node.count );
        wprintf ( L"    [deliver] %s received \"%s\" (src=%s%s)\n",
                  node.name, m.text(), m.source, m.broadcast ? L", broadcast" : L"" );
        fflush ( stdout );
        node.gate.open();
    });
    hub.onError ( [&node] ( const wchar_t *what )
    {
        wprintf ( L"    [error]   at %s: %s\n", node.name, what );
        fflush ( stdout );
    });
}

static bool Check ( const wchar_t *label, bool ok )
{
    wprintf ( L"    => %s: %s\n\n", ok ? L"PASS" : L"FAIL", label );
    fflush ( stdout );
    return ok;
}

int main ( )
{
    light::InitConsole();
    wprintf ( L"=== RouteLoopbackTest (Light) - four hubs, one process, tree routing ===\n" );
    wprintf ( L"        %s\n       /        \\\n    %s   %s\n      |\n   %s\n\n",
              kRoot, kA, kB, kLeaf );
    fflush ( stdout );

    Node nRoot ( kRoot ), nA ( kA ), nB ( kB ), nLeaf ( kLeaf );

    try
    {
        p2pf::Network net;

        // ---- Build the tree. Listeners first: dmx:// dials do not retry. ----
        //
        // Every edge here joins a parent to a child, which is what makes the
        // routing this harness measures possible at all: the kernel's onward
        // rules require an ancestor/descendant relation, so a sibling edge
        // (Root.A <-> Root.B, say) would carry direct traffic and still refuse
        // to be a transit hop. The arming verbs classify each pair and would
        // say so -- every call below returns a plain S_OK.
        const std::wstring strSvcRootA = light::Dmx ( kSvcRootA );
        const std::wstring strSvcRootB = light::Dmx ( kSvcRootB );
        const std::wstring strSvcALeaf = light::Dmx ( kSvcALeaf );

        p2pf::Hub root = net.createHub ( kRoot );
        Attach ( root, nRoot );
        HRESULT hr = root.listen ( kA, strSvcRootA.c_str() );
        if ( SUCCEEDED(hr) ) hr = root.listen ( kB, strSvcRootB.c_str() );
        if ( FAILED(hr) )
        {
            light::Log ( L"MAIN", L"FATAL: root Dmx listen failed (%s)", light::HrName ( hr ) );
            return light::EXIT_SETUP;
        }

        p2pf::Hub a = net.createHub ( kA );
        Attach ( a, nA );
        hr = a.listen ( kLeaf, strSvcALeaf.c_str() );   // arm before Leaf dials
        if ( SUCCEEDED(hr) ) hr = a.connect ( kRoot, strSvcRootA.c_str() );
        if ( FAILED(hr) )
        {
            light::Log ( L"MAIN", L"FATAL: A wiring failed (%s)", light::HrName ( hr ) );
            return light::EXIT_SETUP;
        }

        p2pf::Hub b = net.createHub ( kB );
        Attach ( b, nB );
        hr = b.connect ( kRoot, strSvcRootB.c_str() );
        if ( FAILED(hr) )
        {
            light::Log ( L"MAIN", L"FATAL: B wiring failed (%s)", light::HrName ( hr ) );
            return light::EXIT_SETUP;
        }

        p2pf::Hub leaf = net.createHub ( kLeaf );
        Attach ( leaf, nLeaf );
        hr = leaf.connect ( kA, strSvcALeaf.c_str() );
        if ( FAILED(hr) )
        {
            light::Log ( L"MAIN", L"FATAL: Leaf wiring failed (%s)", light::HrName ( hr ) );
            return light::EXIT_SETUP;
        }

        // Give the three links their login handshakes.
        ::Sleep ( 1500 );
        wprintf ( L"Links: Root<->A %s   Root<->B %s   A<->Leaf %s\n\n",
                  root.isPeerUp ( kA )   ? L"up" : L"DOWN",
                  root.isPeerUp ( kB )   ? L"up" : L"DOWN",
                  a.isPeerUp    ( kLeaf ) ? L"up" : L"DOWN" );
        fflush ( stdout );

        bool allOk = true;

        // ---- [1] Non-adjacent, up then down: Leaf -> B ----------------------
        wprintf ( L"[1] Leaf -> %s  (non-adjacent: up to the root, then down)\n", kB );
        fflush ( stdout );
        hr = leaf.sendText ( kB, kTopic, L"hello from the leaf" );
        wprintf ( L"    sendText returned %s\n", light::HrName ( hr ) );
        bool got1 = nB.gate.wait ( 5000 );
        allOk &= Check ( L"Leaf -> B delivered across two hops", got1 && nB.count == 1 );

        // ---- [2] Straight down: Root -> Leaf --------------------------------
        wprintf ( L"[2] Root -> %s  (straight down the tree)\n", kLeaf );
        fflush ( stdout );
        hr = root.sendText ( kLeaf, kTopic, L"ping, down we go" );
        wprintf ( L"    sendText returned %s\n", light::HrName ( hr ) );
        bool got2 = nLeaf.gate.wait ( 5000 );
        allOk &= Check ( L"Root -> Leaf delivered via A", got2 && nLeaf.count == 1 );

        // ---- [3] No such address --------------------------------------------
        wprintf ( L"[3] Leaf -> %s  (unknown branch; must not be delivered)\n", kZ );
        fflush ( stdout );
        LONG before = nRoot.count + nA.count + nB.count + nLeaf.count;
        hr = leaf.sendText ( kZ, kTopic, L"lost, no such node" );
        wprintf ( L"    sendText returned %s\n", light::HrName ( hr ) );
        ::Sleep ( 1500 );
        LONG after = nRoot.count + nA.count + nB.count + nLeaf.count;
        allOk &= Check ( L"Leaf -> Z delivered to nobody", before == after );

        // ---- [4] Broadcast relayed down the tree ----------------------------
        wprintf ( L"[4] Root broadcast  (relayed to every descendant -- no counterpart\n"
                  L"    in the original, this is the tree behaviour MSCS adds)\n" );
        fflush ( stdout );
        LONG aBefore = nA.count, bBefore = nB.count, leafBefore = nLeaf.count;
        static const wchar_t kAllHands[] = L"all hands";
        hr = root.broadcast ( kTopic, kAllHands, (unsigned int)sizeof(kAllHands) );
        wprintf ( L"    broadcast returned %s\n", light::HrName ( hr ) );
        ::Sleep ( 2000 );
        wprintf ( L"    reached: A %s   B %s   Leaf %s\n",
                  nA.count    > aBefore    ? L"yes" : L"no",
                  nB.count    > bBefore    ? L"yes" : L"no",
                  nLeaf.count > leafBefore ? L"yes" : L"no" );
        allOk &= Check ( L"broadcast reached both direct children",
                         nA.count > aBefore && nB.count > bBefore );

        light::Log ( L"MAIN", L"shutdown begin" );
        return light::Verdict ( allOk,
                                L"the tree routed every scenario as expected",
                                L"a route did not match expectations" );
    }
    catch ( const std::exception& e )
    {
        printf ( "FATAL: %s\n", e.what() );
        return light::EXIT_SETUP;
    }
}
